### ISSUE TITLE
Build Js multithreaded seems to fail

### ISSUE THREAD

**Issue by momo-AUX1** [danoon2/Boxedwine#107]
Build JS multithreaded seems to fail due to undefined because it's looking for undefined does anyone know why? 

Built with the latest EMScripten as of today with mingw on windows 10

**Comment by danoon2** [comment]
@momo-AUX1 I probably shouldn't have checked in that build script or maybe I should have put a note on it.  It's just an experiment.  I did get it to work on a slightly older version of emscripten, but it didn't run for more than a few seconds.  In my proof of concept it wasn't any faster, but I have some ideas of how to improve the code.

We need a newer version of emscripten to work with multi-threaded, but the newer versions of emscripten broke file handling.  My goal is to get Boxedwine working on the newest Emscripten, but I don't have a time frame for that.  It would be nice if I can get it done and make my next release of Boxedwine in January for Wine 10 release.

Here is my experiment of multi-threaded using Emscripten 3.1.61

https://www.boxedwine.org/v/mt/kevin61mt/boxedwine.html?app=mdk&p=PERF_W95.EXE

**Comment by momo-AUX1** [comment]
Amazing work! I recently ported this specific EMScripten app to Xbox what are some general optimization flags do you recommend? 

As for the multithreaded version, I'll keep updating it and make sure it matches the latest BoxedWine release it's currently at Debian10+wine5.0 

Thank you very much for your hard work!

**Comment by danoon2** [comment]
@momo-AUX1 Do you mean the emscripten build is running on xbox instead of the native c++ version?

**Comment by momo-AUX1** [comment]
Indeed the native version would've been a huge task to get it working, but the EMScripten version with some tweaks got up and running! That's why I'm hoping we'd see a multithreaded JS build or any optimizations on that front I could send you the app to test if needed

**Comment by danoon2** [comment]
@momo-AUX1 I was curious and just looked into compiling for xbox.  https://wiki.libsdl.org/SDL2/README/gdk  It seems like there are a lot of hoops to jump through.  I was hoping to see just a simple build on windows with an emulator on windows, something more like iPad development.  I guess not.

**Comment by momo-AUX1** [comment]
There is a SDL for UWP (https://github.com/aerisarn/SDL-uwp-gl) although I wouldn't really know what wine would need on top of SDL so I didn't really attempt it sadly.

However I have a question, what are some optimization flags that you recommend for xbox/EMSscripten? And could boxedwine support Linux apps or is it windows only?

**Comment by momo-AUX1** [comment]
Either way, the reception for it on xbox has been stellar you could expect a video about it to be released soon! I'll keep on maintaining it to the latest EMScripten build so the users can benefit with. big disclaimer about performance for now

**Comment by momo-AUX1** [comment]
Amazing news, I'll make sure to rebase the Xbox on this version. Would you happen to know what's the perf gains compared to the old build ?

**Comment by danoon2** [comment]
@momo-AUX1 also, if you are building yourself, you can now use the latest, 4.0 for emscripten.  The multi-threaded normal CPU core is about 50%-100% faster on other platforms and even more importantly, it should feel more responsive with less tiny pauses in the UI.  In my limited testing on emscripten, it looks like this will be the case for emscripten too.

**Comment by momo-AUX1** [comment]
Understood you can expect it done today or tomorrow I will test all the new improvements thoroughly on Xbox and report them

**Comment by momo-AUX1** [comment]
Looks like it's failing to read my zip file which otherwise worked back then it says this :

```
boxedwine.zip depends on debian10.zip, and debian10.zip could not be found
Using zip file system: boxedwine.zip
Using zip file system: boxedwine.zip/debian10.zip
Loaded boxedwine.zip in 5332 ms
Could not load zip file: boxedwine.zip/debian10.zip
Could not read file global info from zip file: boxedwine.zip/debian10.zip
Loaded boxedwine.zip/debian10.zip in 29 ms
wine: Unhandled page fault on read access to FFFFFFFF at address D120B126 (thread 0009), starting debugger...
boxedwine-shell.js:613 Can't attach process 0008: error 5
boxedwine-shell.js:613 winedbg: Internal crash at 0xd124b126
```

I also do see it uses significantly more memory and the threads are there just hoping it works on xbox, the new header will need some solving. 

If you have a spare FS zip I could test to confirm it isn't a zip issue I'll be glad to try it out

**Comment by momo-AUX1** [comment]
@kevodwyer looks like it was indeed a zip issue your build artifact works perfectly, could you please link me to a guide on how to make a full debian10 system again? I seriously don't know how the zip suddenly started failing after it worked

**Comment by momo-AUX1** [comment]
boxedwine-shell.js:625 
 TypeError: Cannot read properties of undefined (reading 'audioContext')

Looks like it fails to load audio for the game I tested which was paintball from the XP classics

Also the blue background is behind the window is that intentional ? 

I'm guessing it's because it's a beta build but so far performance is really great

**Comment by danoon2** [comment]
@momo-AUX1 The full Wine 6.0 file system can be downloaded from here

http://www.boxedwine.org/v2/2/TinyCore15Wine6.0.zip

the paintball game is not full screen, so it's normal that there would be a blue desktop behind it.  But if it looks odd, can you post a screen shot?

**Comment by momo-AUX1** [comment]
![Image](https://github.com/user-attachments/assets/77334bc0-084e-4e17-b0af-d672265e6c32)



as you see the desktop background is white and the background is blue? even though I forced it to be 100vh and 100vw the css code that works correctly resizes the main window but the background still stays for some reason can at least say it's specific to the code and not the OS since 3 zips and the same issues are present

**Comment by danoon2** [comment]
@momo-AUX1 It looks like the explorer shell is not taking up the entire screen.  You can tell by the gray bottom bar not going all the way across.

If you know the resolution you will need, you can tell the shell what size to be by launch it with something like

/bin/wine explorer /desktop=shell,1024x786

**Comment by momo-AUX1** [comment]
Just tried this:

```bash
http://localhost:8000/boxedwine.html?resolution=1280x720&app=pinball.zip


Command line arguments: "./this.program" "-root" "/root" "-zip" "TinyCore15Wine6.0.zip" "-mount" "pinball.zip" "/home/username/.wine/dosdevices/c:/files" "-mount_drive" "/d_drive" "d" "-resolution" "1280x720" "/bin/wine" "explorer" "/desktop=shell, 1280x720"
```


Doesn't seem to change anything I'll try compiling a non MT version and see if this still persists

**Comment by danoon2** [comment]
@momo-AUX1 Last thing I can think of, you should change

"/desktop=shell, 1280x720"

to

"/desktop=shell,1280x720"

no space after comma

**Comment by momo-AUX1** [comment]
becomes just a white page sadly

**Comment by momo-AUX1** [comment]
I don't know if you guys saw this but will you consider using it to maybe run GL on EMS? 

https://github.com/ptitSeb/gl4es

this project uses it in WASM to run quake

https://github.com/GMH-Code/Qwasm

**Comment by danoon2** [comment]
It seems like running OpenGL on Emscripten is hit or miss.  I get the feeling the lower the OpenGL version the better.  Quake 1 and Quake 2 both used a pretty low version of OpenGL.  I think Quake 1 used OpenGL 1.1, with GLQuake using a subset of that called MiniGL.  Interestingly, 3dfx had an OpenGL driver for their card, but it wasn't complete, it seemed like it implemented just enough to run Quake.

On Emscripten's page, https://emscripten.org/docs/porting/multimedia_and_graphics/OpenGL-support.html, it states "support for a number of legacy GL 1.x features and commands."

So without a wrapper, it might be possible to run some OpenGL 1.x stuff in Emscripten.

Emscripten: "OpenGL ES 2.0/3.0 emulation support for some emulated OpenGL ES 2.0/3.0 features that are not present in WebGL."

So with an ES wrapper, it might be possible to get farther.

As far as running Boxedwine with 3D support in Emscripten?  I have been thinking of this lately and haven't had much luck.  In Wine 3.0, https://www.winehq.org/announce/3.0, they stated

Android:

- OpenGL is supported, but it's limited to the OpenGL ES API that is
  available on Android.

- Direct3D is not supported yet, because it cannot run on top of OpenGL ES
  at this point. This will be addressed during the next development cycle.

But then nothing, I don't see android and Direct3D mentioned again.  So I guess it wasn't a priority or it is just a really difficult problem.  Around that time they pivoted to getting Direct3D on Vulkan to run.

At this point I was hoping to find a direct 3d 1-8 wrapper that converted to OpenGL ES, but I don't think that ever existed.  The closest I got was https://dxgl.org/, which can run directx 1-7 on OpenGL 2.0.  But I haven't had much luck running that in Boxedwine.  It crashes an OpenGL call, I'm not sure if that's my fault or not.  I may look into that more, but I don't have much hope that OpenGL 2.0 would run in Emscripten.

So, while Quake running GL in Emscripten is interesting, I think its the easiest case possible because Quake doesn't use the full OpenGL 1.1 spec and its such an old version of OpenGL.  Getting DirectX 1-8 should be easier that DirectX 9, but so far I haven't had any luck.  

I think software emulation might be the way to go.  Swiftshader sounds promising

https://github.com/ddrmax/swiftshader-ex

But compiling this for win32 and running in Boxedwine will probably be pretty slow.

**Comment by momo-AUX1** [comment]
Yeah it might be slower but great for compatibility I'm hoping browsers or devices will get better to eventually compensate for the software renderer.

**Comment by danoon2** [comment]
@momo-AUX1 I agree.   I got Mesa LLVM software OpenGL to run on Arm Mac and win32/win64/arm64 Windows.  For the older games, it runs pretty well.
